plugins/generic/recommendBySimilarity/RecommendBySimilarityPlugin.inc.php

```diff
@@ -56,45 +62,87 @@ function getDescription() {
 	//
 	/**
 	 * @see templates/article/footer.tpl
+	 * @todo Revert back to using the ArticleSearch::retrieveResults() once its performance is addressed
 	 */
-	function callbackTemplateArticlePageFooter($hookName, $params) {
-		$smarty =& $params[1];
+	function callbackTemplateArticlePageFooter(string $hookName, $params) {
+		/** @var Smarty */
+		$smarty = $params[1];
 		$output =& $params[2];
 
 		// Identify similarity terms for the given article.
 		$displayedArticle = $smarty->getTemplateVars('article');
 		$articleId = $displayedArticle->getId();
 		import('classes.search.ArticleSearch');
+		import('classes.search.ArticleSearchIndex');
 		$articleSearch = new ArticleSearch();
-		$searchTerms = $articleSearch->getSimilarityTerms($articleId);
-		if (empty($searchTerms)) return false;
+		$keywords = (new ArticleSearchIndex())->filterKeywords($articleSearch->getSimilarityTerms($articleId));
+		$keywords = array_filter(array_unique($keywords), 'strlen');
+		if (!count($keywords)) {
+			return false;
+		}
 
-		// If we got similarity terms then execute a search with...
-		// ... request, journal and error messages, ...
 		$request = Application::get()->getRequest();
 		$router = $request->getRouter();
 		$journal = $router->getContext($request);
-		$error = null;
-		// ... search keywords ...
-		$query = implode(' ', $searchTerms);
-		$keywords = array(null => $query);
-		// ... and pagination.
 		$rangeInfo = Handler::getRangeInfo($request, 'articlesBySimilarity');
-		$rangeInfo->setCount(RECOMMEND_BY_SIMILARITY_PLUGIN_COUNT);
-		$smarty->assign(array(
-			'articlesBySimilarity' => $articleSearch->retrieveResults(
-					$request,
-					$journal,
-					$keywords,
-					$error,
-					null, null,
-					$rangeInfo,
-					array($articleId)
-			),
-			'articlesBySimilarityQuery' => $query,
-		));
+		$rangeInfo->setCount(static::RECOMMEND_BY_SIMILARITY_PLUGIN_COUNT);
+
+		/** @var Builder */
+		$queryBuilder = (new SubmissionQueryBuilder())
+			->filterByContext($journal->getId())
+			->filterByStatus(STATUS_PUBLISHED)
+			->getQuery();
+
+		$queryBuilder->where('s.submission_id', '<>', $articleId);
+
+		$keywordIdFields = [];
```

Comment:
I'll add comments/prettify it a bit, but before doing that, I wanted to know if the approach is ok.

I'm basically replacing the `ArticleSearch::retrieveResults()`, and while doing that, I've tried to keep the "ranking".
1. Number of keywords matched (A record with 1 `keywordA` + 1 `keywordB" has higher priority than a record with 100 `keywordA`)
2. Total number of keyword matches

Reply:
In fact, I'd like to drop the ranking for extra performance, perhaps I can add it as a plugin option...

Then as a last improvement, which I'm not going to cover here, would be to add some <script> magic to avoid a full page reload.